src/wx/widgets/sdljoy.cpp

```diff
@@ -38,420 +39,489 @@ static int16_t axisval(int16_t x)
 
-            js_dev = SDL_GameControllerGetJoystick(dev);
+            got_event = true;
+            break;
         }
-    }
-    else {
-        if ((js_dev = SDL_JoystickOpen(joy)))
-            joystate[joy].dev = js_dev;
-    }
 
-    if (js_dev) {
-        auto instance = SDL_JoystickInstanceID(js_dev);
+        case SDL_JOYDEVICEADDED:
+        {
+            int sdl_index = e.jdevice.which;
+            // Always remap all controllers.
+            RemapControllers();
+            got_event = true;
+            break;
+        }
 
-        instance_map[instance] = &(joystate[joy]);
+        case SDL_JOYDEVICEREMOVED:
+        {
+            SDL_JoystickID joy_id = e.jdevice.which;
+            if (add_all_joysticks_)
+                RemapControllers();
+            else
+                DisconnectController(joy_id);
+            got_event = true;
+            break;
+        }
 
-        joystate[joy].instance = instance;
+        default:
+            // Ignore all other events.
+            break;
+        }
     }
 
-    joystate[joy].index = joy;
-    joystate[joy].is_gc = is_gc;
-}
-
-void wxSDLJoy::RemapControllers()
-{
-    for (auto&& joy : joystate) {
-        auto& state = joy.second;
+    bool do_poll = false;
+    wxLongLong now = wxGetUTCTimeMillis();
 
-        DisconnectController(state);
-        ConnectController(joy.first);
+    if (got_event) {
+        last_poll_ = now;
+    } else if (now - last_poll_ > kPollTimeInterval) {
+        do_poll = true;
+        last_poll_ = now;
     }
-}
 
-void wxSDLJoy::DisconnectController(wxSDLJoyState& state)
-{
-    if (auto& dev = state.dev) {
-        if (state.is_gc) {
-            SDL_GameControllerClose(dev);
-        }
-        else {
-            SDL_JoystickClose(dev);
+    if (do_poll) {
+        for (auto&& joy_state : joystick_states_) {
+            joy_state.second->Poll();
         }
-
-        dev = nullptr;
     }
-
-    instance_map.erase(state.instance);
 }
 
-void wxSDLJoy::Add(int8_t joy_n)
-{
-    if (joy_n < 0) {
-        for (uint8_t joy : range(0, SDL_NumJoysticks()))
-            ConnectController(joy);
-
-        add_all = true;
+void wxSDLJoy::DisconnectController(SDL_JoystickID joystick_index) {
+    assert(!add_all_joysticks_);
 
+    auto iter = joystick_states_.find(joystick_index);
+    if (iter == joystick_states_.end()) {
+        // Nothing to do if the joystick was not active.
         return;
     }
 
-    ConnectController(joy_n);
+    // Update needed SDL indexes.
+    int sdl_index = iter->second->player_index() - 1;
+    needed_sdl_indexes_.insert(sdl_index);
+    joystick_states_.erase(iter);
 }
 
-void wxSDLJoy::Remove(int8_t joy_n)
-{
-    add_all = false;
-
-    if (joy_n < 0) {
-        for (auto&& joy : joystate)
-            DisconnectController(joy.second);
-
-        joystate.clear();
-
-        return;
+void wxSDLJoy::RemapControllers() {
+    if (!add_all_joysticks_) {
+        // Save the connected SDL indexes first.
```

Comment:
We shouldn't do this, the wanted player indices should only be set by the API client.

Reply:
This was from an older commit when I was not reconnecting all joysticks on a new connection, which led to some issues. It no longer has any purpose so I removed it.